mithril-aggregator/src/artifact_builder/cardano_database_artifacts/ancillary.rs

```diff
@@ -27,18 +39,104 @@ impl AncillaryFileUploader for LocalUploader {
     }
 
-    pub async fn upload_archive(&self, db_directory: &Path) -> StdResult<Vec<AncillaryLocation>> {
+    /// Returns the list of files and directories to include in the snapshot.
+    /// The immutable file included in the ancillary archive corresponds to the last one (and not finalized yet)
+    /// when the immutable file number given to the function corresponds to the penultimate.
+    fn get_files_and_directories_to_snapshot(immutable_file_number: u64) -> Vec<PathBuf> {
+        let next_immutable_file_number = immutable_file_number + 1;
+        let chunk_filename = format!("{:05}.chunk", next_immutable_file_number);
+        let primary_filename = format!("{:05}.primary", next_immutable_file_number);
+        let secondary_filename = format!("{:05}.secondary", next_immutable_file_number);
+
+        vec![
+            PathBuf::from(VOLATILE_DIR),
+            PathBuf::from(LEDGER_DIR),
+            PathBuf::from(IMMUTABLE_DIR).join(chunk_filename),
+            PathBuf::from(IMMUTABLE_DIR).join(primary_filename),
+            PathBuf::from(IMMUTABLE_DIR).join(secondary_filename),
+        ]
+    }
+
+    /// Creates an archive for the Cardano database ancillary files for the given immutable file number.
+    fn create_ancillary_archive(&self, beacon: &CardanoDbBeacon) -> StdResult<OngoingSnapshot> {
+        debug!(
+            self.logger,
+            "Creating ancillary archive for immutable file number: {}",
+            beacon.immutable_file_number
+        );
+
+        let paths_to_include =
+            Self::get_files_and_directories_to_snapshot(beacon.immutable_file_number);
+
+        let archive_name = format!(
+            "{}-e{}-i{}.ancillary.{}",
+            self.cardano_network,
+            *beacon.epoch,
+            beacon.immutable_file_number,
+            self.compression_algorithm.tar_file_extension()
+        );
+
+        let snapshot = self
+            .snapshotter
+            .snapshot_subset(&archive_name, paths_to_include)
+            .with_context(|| {
+                format!(
+                    "Failed to create ancillary archive for immutable file number: {}",
+                    beacon.immutable_file_number
+                )
+            })?;
+
+        debug!(
+            self.logger,
+            "Ancillary archive created at path: {:?}",
+            snapshot.get_file_path()
+        );
+
+        Ok(snapshot)
+    }
+
+    /// Uploads the ancillary archive and returns the locations of the uploaded files.
+    async fn upload_ancillary_archive(
+        &self,
+        archive_filepath: &Path,
+    ) -> StdResult<Vec<AncillaryLocation>> {
         let mut locations = Vec::new();
         for uploader in &self.uploaders {
-            // TODO: Temporary preparation work, `db_directory` is used as the ancillary archive path for now.
-            let location = uploader.upload(db_directory).await?;
+            let location = uploader.upload(archive_filepath).await?;
```

Comment:
We probably need to:
- check that at least one location is available (or return an error)
- log errors of `upload` calls instead of returning right away